src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,113 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_ARRAY_SIZE = 10;
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE;
+    private int size;
+    private Object[] array;
+
+    ArrayList() {
+        array = new Object[INITIAL_ARRAY_SIZE];
+    }
+
+    private void grow(int value) {
+        if (size + value > MAX_ARRAY_SIZE) {
+            throw new OutOfMemoryError();
+        }
+        Object[] temp = new Object[size + value];
+        System.arraycopy(array, 0, temp, 0, size);
+        array = temp;
+    }
+
+    private void grow() {
+        grow(INITIAL_ARRAY_SIZE / 2);
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("The array index is out of bounds");
+        }
+    }
+
+    private void checkAddIndex(int index) {
```

Comment:
You don't need this method, cause it is called only one time. Put logic from this method in method where it is called.